File: Vantage/Views/Components/EventCard.swift
import SwiftUI

/// One detected change, rendered with its epistemic labels intact.
///
/// The headline is a CALCULATION — measured arithmetic, with the derivation one
/// tap away. The "why this matters" text is an INTERPRETATION and is badged
/// differently, because it is a judgement about what the figures show rather
/// than a measurement. Neither ever becomes a suggestion to buy or sell.
struct EventCard: View {
    let event: DetectedEventDTO
    /// Shown when the card appears in a feed mixing several securities.
    var symbol: String?
    var isNew: Bool = false

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            header

            ClaimRow(claim: event.headlineClaim)

            if !event.detailLines.isEmpty {
                VStack(alignment: .leading, spacing: 3) {
                    ForEach(Array(event.detailLines.enumerated()), id: \.offset) { _, line in
                        Text(line)
                            .font(.system(.caption, design: .monospaced))
                            .foregroundStyle(.secondary)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
            }

            if event.unusualness > 0 {
                UnusualnessMeter(value: event.unusualness)
            }

            if let context = event.contextClaim {
                Divider()
                ClaimRow(claim: context)
            }

            if !event.sourceURLs.isEmpty {
                ForEach(event.sourceURLs, id: \.self) { url in
                    Link(destination: url) {
                        Label("Open the source document", systemImage: "arrow.up.right.square")
                            .font(.caption)
                    }
                }
            }
        }
        .padding(14)
        .background(Color(.secondarySystemGroupedBackground), in: .rect(cornerRadius: 12))
    }

    private var header: some View {
        HStack(spacing: 8) {
            Image(systemName: event.kind.systemImage)
                .font(.caption)
                .foregroundStyle(.secondary)
            if let symbol {
                Text(symbol)
                    .font(.system(.caption, design: .monospaced).weight(.semibold))
            }
            Text(event.kind.displayName)
                .font(.caption.weight(.medium))
                .foregroundStyle(.secondary)
            if isNew {
                Text("NEW")
                    .font(.system(.caption2, design: .monospaced).weight(.semibold))
                    .padding(.horizontal, 5).padding(.vertical, 2)
                    .background(.tint.opacity(0.15), in: .rect(cornerRadius: 4))
                    .foregroundStyle(.tint)
            }
            if event.isProvisional {
                Text("SESSION OPEN")
                    .font(.system(.caption2, design: .monospaced).weight(.semibold))
                    .padding(.horizontal, 5).padding(.vertical, 2)
                    .background(.orange.opacity(0.15), in: .rect(cornerRadius: 4))
                    .foregroundStyle(.orange)
                    .accessibilityLabel("Session still open. This reading can "
                        + "change before the close and is not yet recorded.")
            }
            Spacer()
            Text(Format.dayAndMonth(event.occurredAt))
                .font(.caption)
                .foregroundStyle(.tertiary)
                .monospacedDigit()
        }
    }
}

/// How unusual a reading is relative to the security's own recent history.
///
/// Labelled in words as well as drawn, and deliberately not tinted red or
/// green: unusual is not the same as bad, and this figure carries no direction.
/// It is a position within an observed sample, never a probability — a bar that
/// implied "2% chance" would be a fabricated statistic.
struct UnusualnessMeter: View {
    /// 0–1.
    let value: Double

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack {
                Text("Unusual for this security")
                    .font(.caption).foregroundStyle(.secondary)
                Spacer()
                Text(descriptor)
                    .font(.caption.weight(.medium))
                    .foregroundStyle(.secondary)
            }
            GeometryReader { proxy in
                ZStack(alignment: .leading) {
                    Capsule().fill(.quaternary)
                    Capsule()
                        .fill(.secondary)
                        .frame(width: max(2, proxy.size.width * min(max(value, 0), 1)))
                }
            }
            .frame(height: 4)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Unusual for this security: \(descriptor). "
            + "Measured against its own recent history. Not a measure of "
            + "importance, direction, or likelihood.")
    }

    private var descriptor: String {
        switch value {
        case ..<0.5: "Within its usual range"
        case ..<0.9: "Above its usual range"
        case ..<0.99: "Rare for this security"
        default: "Among the most extreme in the sample"
        }
    }
}

#Preview {
    ScrollView {
        VStack(spacing: 12) {
            EventCard(
                event: DetectedEventDTO(
                    kind: .unusualVolume,
                    occurredAt: .now,
                    headline: "Volume 2.4× the 60-session median on Aug 26",
                    detailLines: [
                        "Volume: 214.9M shares",
                        "Median of prior 60 sessions: 89.5M shares",
                        "Higher than 60 of those 60 sessions"
                    ],
                    context: "Elevated volume means more shares changed hands than usual, "
                        + "which indicates unusual attention. It carries no direction on its own.",
                    unusualness: 1.0,
                    sourceDetails: ["Daily bars"]
                ),
                symbol: "NVDA",
                isNew: true
            )
        }
        .padding()
    }
    .background(Color(.systemGroupedBackground))
}

File: app/src/commonMain/kotlin/com/tylerabitbol/libra/ui/components/EventCard.kt
package com.tylerabitbol.libra.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.tylerabitbol.libra.models.core.DetectedEventDTO
import com.tylerabitbol.libra.support.Format
import com.tylerabitbol.libra.ui.LibraAlpha
import com.tylerabitbol.libra.ui.LibraShapes
import com.tylerabitbol.libra.ui.LibraTheme
import com.tylerabitbol.libra.ui.LibraType
import kotlin.math.max
import kotlin.math.min

/**
 * One detected change, rendered with its epistemic labels intact.
 *
 * The headline is a CALCULATION — measured arithmetic, with the derivation one
 * tap away. The "why this matters" text is an INTERPRETATION and is badged
 * differently, because it is a judgement about what the figures show rather
 * than a measurement. Neither ever becomes a suggestion to buy or sell.
 */
@Composable
fun EventCard(
    event: DetectedEventDTO,
    modifier: Modifier = Modifier,
    /** Shown when the card appears in a feed mixing several securities. */
    symbol: String? = null,
    isNew: Boolean = false,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .libraCard(),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        EventHeader(event, symbol, isNew)

        ClaimRow(event.headlineClaim)

        if (event.detailLines.isNotEmpty()) {
            Column(verticalArrangement = Arrangement.spacedBy(3.dp)) {
                for (line in event.detailLines) {
                    Text(line, style = LibraType.figure, color = LibraTheme.colors.secondaryText)
                }
            }
        }

        if (event.unusualness > 0) {
            UnusualnessMeter(event.unusualness)
        }

        event.contextClaim?.let {
            HorizontalDivider()
            ClaimRow(it)
        }
    }
}

@Composable
private fun EventHeader(event: DetectedEventDTO, symbol: String?, isNew: Boolean) {
    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Swift leads with `event.kind.systemImage`. SF Symbols has no
        // multiplatform counterpart, and the kind is already named in words
        // beside it, so the glyph is dropped rather than redrawn eighteen times
        // for decoration.
        if (symbol != null) {
            Text(
                symbol,
                style = LibraType.tickerSmall,
            )
        }
        Text(
            event.kind.displayName,
            style = MaterialTheme.typography.bodySmall.copy(fontWeight = FontWeight.Medium),
            color = LibraTheme.colors.secondaryText,
            modifier = Modifier.weight(1f, fill = false),
        )
        if (isNew) {
            Tag("NEW", MaterialTheme.colorScheme.primary)
        }
        if (event.isProvisional) {
            Tag(
                "SESSION OPEN",
                LibraTheme.colors.caution,
                accessibilityLabel = "Session still open. This reading can change " +
                    "before the close and is not yet recorded.",
            )
        }
        Text(
            Format.dayAndMonth(event.occurredAt),
            style = LibraType.figure,
            color = LibraTheme.colors.tertiaryText,
            modifier = Modifier.weight(1f, fill = false),
        )
    }
}

@Composable
private fun Tag(text: String, tint: Color, accessibilityLabel: String? = null) {
    Text(
        text,
        style = LibraType.codeSmallEmphasis,
        color = tint,
        modifier = Modifier
            .libraChip(tint)
            .then(
                if (accessibilityLabel == null) {
                    Modifier
                } else {
                    Modifier.semantics { contentDescription = accessibilityLabel }
                },
            ),
    )
}

/**
 * How unusual a reading is relative to the security's own recent history.
 *
 * Labelled in words as well as drawn, and deliberately not tinted red or green:
 * unusual is not the same as bad, and this figure carries no direction. It is a
 * position within an observed sample, never a probability — a bar that implied
 * "2% chance" would be a fabricated statistic.
 */
@Composable
fun UnusualnessMeter(value: Double, modifier: Modifier = Modifier) {
    val descriptor = unusualnessDescriptor(value)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                contentDescription = "Unusual for this security: $descriptor. " +
                    "Measured against its own recent history. Not a measure of " +
                    "importance, direction, or likelihood."
            },
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(Modifier.fillMaxWidth(), Arrangement.SpaceBetween) {
            Text(
                "Unusual for this security",
                style = MaterialTheme.typography.bodySmall,
                color = LibraTheme.colors.secondaryText,
            )
            Text(
                descriptor,
                style = MaterialTheme.typography.bodySmall.copy(
                    fontWeight = FontWeight.Medium,
                ),
                color = LibraTheme.colors.secondaryText,
            )
        }
        ProportionBar(fraction = min(max(value, 0.0), 1.0).toFloat(), fromEnd = false)
    }
}

internal fun unusualnessDescriptor(value: Double): String = when {
    value < 0.5 -> "Within its usual range"
    value < 0.9 -> "Above its usual range"
    value < 0.99 -> "Rare for this security"
    else -> "Among the most extreme in the sample"
}
